cellmlmanip/_singularity_fixes.py

```diff
@@ -0,0 +1,234 @@
+                    u = (find_U[U_wildcard] + log(find_U[Z]))
+                    try:
+                        # Find the singularity point by solving u for V==0, excluding irratinal results
+                        sp = tuple(filter(lambda s: not s.has(I), solveset(u, V)))
+                        if sp:
+                            # we found a singularity point, now find vs,ve (or vmin, vmax)
+                            # by solving for U_offset either side of U
+                            find_v_low = solveset(u + U_offset, V)
+                            find_v_up = solveset(u - U_offset, V)
+                            find_v_low = tuple(find_v_low)
+                            find_v_up = tuple(find_v_up)
+                            assert find_v_low and find_v_up and len(find_v_low) == len(find_v_up) == len(sp) == 1, \
+                                'Expecting exactly 1 singularity point '
+                            (vs, ve, sp) = (find_v_low[0], find_v_up[0], sp[0])
+                    except TypeError:
+                        pass  # Result could be 'ConditionSet' which is not iterable and not Real
+
+                if vs is not None:  # check top
+                    for n in num:  # Check arguments in numerator (or denominator)
+                        match = n.match(Z * V - Z * SP_wildcard)  # search for a multiple of V - sp
+                        found_on_top = check_top_match(match, sp)
+                        if not found_on_top:
+                            # search for a exp(multiple of V - sp)
+                            match = n.match(exp_function(Z * V - Z * SP_wildcard))
+                            found_on_top = check_top_match(match, sp)
+                        if found_on_top:  # We've found a match stop looking in the other numerator arguments
+                            break
+                    if found_on_top:  # found singularity, no need to try further
+                        break
+                    else:
+                        (vs, ve, sp) = None, None, None
+        if vs is not None and found_on_top:  # found singularity, no need to try further
+            break
+        else:
+            (vs, ve, sp) = None, None, None
+
+    # Put dummies back in and return the singularity point and range bundries
+    return (float_dummies(vs), float_dummies(ve), float_dummies(sp))
+
+
+def _fixe_expr_parts(expr, V, U_offset, exp_function):
+    """Removes fixable singularities and replaces it with a piecewise
+    :return: either (Vmin (vs), Vmax (ve), singularity point, expression, True)
+                    if we have identified a singularity needs to be made but it can't be done yet.
+                    For eample if 2 singularities with the same singularity point are added up,
+                    a single singularity is constructed instead.
+             or `(None, None, None, fixed expr, fixed_expr has piecewise)` otherwise
```

Comment:
You need double \`\` in docstrings

Reply:
Use `make html` in the `docs` subdirectory to check that this all compiles into nice looking html

Reply:
Actually, since all this is hidden we won't be rendering it 😆 

Still need to carefully write these docstrings to explain what the code below is supposed to do though. Otherwise I can't review if it's doing that correctly!